design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+harder w.r.t. the cases mentioned in the Background:
+
+1. The CRDs of the Managed Resources defined as `base` and traversed by
+   `patches` might not yet have been created when the Composition is being
+   validated.
+2. The CRD of the Composite Resources specified by `CompositeTypeRef` and
+   traversed by `patches` might not yet have been created when the Composition
+   is being validated.
+3. We do not have all the required CRDs locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+In order to address the possible absence of Managed Resources' CRDs, Composite
+Resources' CRDs or any other required external resource at validation time we
+should allow explicitly setting one of 2 modes, `strict` and `loose`(default) .
+These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that any missing resource
+should result in a direct rejection or a possibly incomplete validation.
+
+In `loose` mode we should emit warnings in case any of the preconditions for a
+check are not met, and skip further validation related to that check.
+
+Regarding the approach, I propose to implement a Composition Validating Webhook
+by validating each `resource` and final rendered `Composed
+Resource`, querying the API Server to get the needed schemas.
+
+I suggest we avoid validating `bases`. Doing that would require to ignore
+validating required fields, as those could be set at a later stage by patches
+or transformations. Assuming we implement full `patches` type coverage,
+validating `bases` per se can be avoided, as any error should either be caught
+validating the patches or can be attributed to the `base` itself.
+
+To validate `patches`, we will need to implement ad-hoc functions, because the
+current patching code is dealing directly with unstructured data and does not
+take into consideration schemas at all, therefore not reporting any schema
+related error. The code implemented for [upbound/upjet][upject-validation-code]
+or [crossplane-contrib/crossplane-lint][crossplane-lint-validation-code] could
+be good starting points for this.
+
+To render the Composed Resources, we will need to reuse the Composite resources
+reconciliation loop as much as possible, extracting the rendering logic to
+remove any write operation it currently performs. Which is similar to what has
+been done in the LSP [implementation][lsp-validation-code].
```

Comment:
It's not clear to me whether this validation would happen at XR-creation time, or at Composition-creation time. If the latter, does this mean we'll use dummy inputs?

Reply:
Composition creation time. The assumption is that: "A Composition can be considered valid if all patches and transformations are valid per se, and given a minimal mocked XR according to the XRD schema, produces all valid rendered resources according to each resource's schema". See reference implementation for more details: https://github.com/crossplane/crossplane/pull/3902/files#diff-1681f727532a5c3b92fda0de12dd176ef3d9320676691daf5bb5cb0e4d0221b5R78.

So yes, we would need to mock inputs, mostly the XR, but we might think of mocking also other inputs, e.g. EnvironmentConfigs or decide to let the reconciler use actual ones if present, for read only operations. We did not handle EnvironmentConfigs in the reference implementation at the moment, only mocking the minimal required XR.

Reply:
Got it. It wasn't obvious to me reading the design doc that we had decided to mock inputs. I'll re-read today and see whether I just missed it.